### ISSUE TITLE
Can we check for field names that are duplicated when composing types? 

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/chipper#1266]
In https://github.com/phetsims/scenery/issues/1426, I identified some distrubing duplication of options in the `AccessibleValueHandler` stack.  It's dsturbing because `optionize` allowed us to create this mess.  Should `optionize` be checking for duplicate field names?  When would it be acceptable to have duplicate field names?

**Comment by samreid** [comment]
Today @pixelzoom asked about a function that combines types, maybe something in this direction?

```ts
type IntersectTypes<A, B extends { [K in keyof B]: K extends keyof A ? never : B[K] }> = A & B;
```

**Comment by pixelzoom** [comment]
I changed the title of this issue, since it's not relevant to `optionize`.

**Comment by samreid** [comment]
I saw a Without type and Xor type that look like they may be related or share building blocks with what we want to do here: https://stackoverflow.com/questions/42123407/does-typescript-support-mutually-exclusive-types

**Comment by samreid** [comment]
A note that in https://github.com/phetsims/phet-core/issues/108 we discussed a "one stop shop" for combining options.  It isn't fully flexible, but it may have ideas about how to check for duplicate field names.

**Comment by zepumph** [comment]
We want this, but no one has time for it currently. Marking as an epic to be discussed in future quarterly goals.

**Comment by samreid** [comment]
> We want this, but no one has time for it currently. 

The proposal in https://github.com/phetsims/chipper/issues/1266#issuecomment-1158028197 works correctly to make sure there is no overlap in the keys. That pattern can be extended to accept more type parameters:

```ts
declare type NoIntersectionKeysAllow<A,
  B extends { [K in keyof B]: K extends keyof A ? never : B[K] },

  // assumes A and B have something, so we don't end up with {} & {} & {}
  C extends { [K in keyof C]: K extends keyof A ? never : K extends keyof B ? never : C[K] } = {},
  D extends { [K in keyof D]: K extends keyof A ? never : K extends keyof B ? never : K extends keyof C ? never : D[K] } = {}>
  = A & B & C & D;
```

However, using this implementation impacts the readability at usage sites.  For instance:

```ts
export type GOSimOptions = SelfOptions & PickOptional<SimOptions, 'phetioDesigned'>;
```

becomes:
```ts
export type GOSimOptions = NoIntersectionKeysAllow<SelfOptions, PickOptional<SimOptions, 'phetioDesigned'>>;
```

(I would prefer a more succinct name but went with the one mentioned by @zepumph in https://github.com/phetsims/chipper/issues/1266#issuecomment-1164724822 ).

Furthermore, we expose trouble in cases like DefaultSliderTrack which has:

```ts
export type DefaultSliderTrackOptions = SelfOptions & SliderTrackOptions & PickRequired<SliderTrackOptions, 'size' | 'enabledRangeProperty'>;
```

Since `size` flows in from `SliderTrackOptions` and the `PickRequired` part, `NoIntersectionKeysAllow` would prevent sharing keys.  We could StrictOmit those keys from the SliderTrackOptions.  But I suspect it would be better to create a more readable API for this as described in https://github.com/phetsims/phet-core/issues/108 and https://github.com/phetsims/phet-info/issues/187#issuecomment-1059807464.  One of the disadvantages there was that we didn't know how to annotate the required and optional parts, so they were easy to confuse.  Here is a pattern with the following features:

* Allows us to annotate what should be required vs optional
* Forbids us from marking an optional thing as optional, or a required thing as required.

If we like this pattern, we can extend it so that we can omit `required` or `optional` if there aren't any, instead of having to say `never`:

```ts
// https://stackoverflow.com/questions/54520676/in-typescript-how-to-get-the-keys-of-an-object-type-whose-values-are-of-a-given
// https://stackoverflow.com/questions/52984808/is-there-a-way-to-get-all-required-properties-of-a-typescript-object
type RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

type ReqOpt<A, B extends {
  required: OptionalKeys<A>;
  optional: RequiredKeys<A>;
}> = PickRequired<A, B['required']> & PickOptional<A, B['optional']>;

export type DefaultSliderTrackOptions = ReqOpt<NoIntersectionKeysAllow<SelfOptions, SliderTrackOptions>, {
  required: 'size';
  optional: never;
}>;
```

In summary, a type that forbids overlapping keys is easy.  The time consuming parts for this issue will be:

* Deciding on the API that we want for combining options types (must support required/optional, and we have to decide whether it is a type error if we mark something as optional but it was already optional), and we need to address the "sparse" form where you pick a few things vs the "rich" form where you adopt everything.
* Implementing the proposed API
* Visiting all usage sites and using the implementation
* Lint rules to keep us using the desired pattern.  An autofix could help address the prior bullet point.

**Comment by chrisklus** [comment]
From 8/18/22 dev meeting:

We ran out of time for this and will bring it back next week when @pixelzoom is back.

**Comment by samreid** [comment]
At today's meeting, we also discussed that this would probably be opt-in on demand, not something we would recommend or advocate everywhere.  Subteam will build an implementation that we can use on a case by case basis.

**Comment by jonathanolson** [comment]
PSA for developers: We've added `ExclusiveIntersection<A, B, [C], [D]>` (and helper types like `WithRequired<A, keys>`, `WithOptional<A, keys>`, so this can be used to ensure there is no overlap of keys.

The normal pattern would be to turn:

```js
type ParentOptions = AOptions & BOptions;
```
to
```js
type ParentOptions = ExclusiveIntersection<AOptions, BOptions>;
```

We've also found a case that has some tradeoffs, and we're curious what's best. A good example where overriding of an option makes sense is in AccessibleValueHandler's `tagName` option. If we create a type `ExclusiveIntersection` that will intersect types and ensure they don't include the same options, we'd have to always `StrictOmit` out `tagName` in every place AccessibleSliderOptions/AccessibleNumberSpinner is used, and that's less-than-ideal:

```js
type ParentOptions = ExclusiveIntersection<AccessibleSliderOptions, StrictOmit<NodeOptions, 'tagName'>>;
```

Compare this to:

```js
type ParentOptions = AccessibleSliderOptions & NodeOptions;
```


however if you're using something like PickRequired to make a certain subset required, that needs to be still done with `&`, e.g.:
```js
type ParentOptions = ExclusiveIntersection<A, B & PickRequired<B, 'tandem'>>
```
or
```js
type ParentOptions = ExclusiveIntersection<A, StrictOmit<B, 'tandem'>, PickRequired<B, 'tandem'>>
```
or
```js
type ParentOptions = ExclusiveIntersection<A, WithRequired<B, 'tandem'>>
```

We've introduced `WithRequired` which is a shortcut for making SOME properties required, while leaving the rest in the same state, and `WithOptional` which does the same thing but making SOME properties optional. Thus the common `A & PickRequired<A, keys>` can be turned into `WithRequired<A, keys>`.

**Comment by zepumph** [comment]
PSA occurred during dev meeting.

So it looks like the pattern like `NodeOptions & PickRequired<NodeOptions, 'tandem'>` is deprecated, and instead we will us `WithRequired`. In the future we will most likely want to change these usages.

We will try these types out and see how we like them!
